Extend Fetch to accept local connections

igower opened the pull request:
#### d4e2bd45ea23cd556ef21a9ae002ac277245a7e8
<pre>
Extend Fetch to accept local connections
<a href="https://bugs.webkit.org/show_bug.cgi?id=296710">https://bugs.webkit.org/show_bug.cgi?id=296710</a>
<a href="https://rdar.apple.com/154439024">rdar://154439024</a>

Reviewed by Alex Christensen.

Integration with fetch as per the local network access spec: (<a href="https://wicg.github.io/local-network-access/#integration-with-fetch)">https://wicg.github.io/local-network-access/#integration-with-fetch)</a>
Upstream commit: <a href="https://github.com/web-platform-tests/wpt/commit/c4ab993576ca17eafd7108853e112d1e97aeaacc">https://github.com/web-platform-tests/wpt/commit/c4ab993576ca17eafd7108853e112d1e97aeaacc</a>

* LayoutTests/TestExpectations:
* LayoutTests/ipc/invalid-url-network-data-task-crash.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::updateTargetAddressSpaceIfNeeded):
(WebCore::FetchRequest::initializeWith):
(WebCore::FetchRequest::resourceRequest const):
(WebCore::FetchRequest::clone):
* Source/WebCore/Modules/fetch/FetchRequest.h:
* Source/WebCore/Modules/fetch/FetchRequest.idl:
* Source/WebCore/Modules/fetch/FetchRequestInit.h:
(WebCore::FetchRequestInit::hasMembers const):
* Source/WebCore/Modules/fetch/FetchRequestInit.idl:
* Source/WebCore/Modules/fetch/IPAddressSpace.cpp: Added.
(WebCore::determineIPAddressSpace):
(WebCore::isLocalIPAddressSpace):
* Source/WebCore/Modules/fetch/IPAddressSpace.h:
* Source/WebCore/Modules/mediasession/MediaSessionCoordinator.cpp:
* Source/WebCore/Modules/mediastream/MediaStreamTrackProcessor.cpp:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpSenderBackend.h:
* Source/WebCore/Modules/notifications/NotificationEvent.cpp:
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/Modules/webaudio/AudioNode.h:
* Source/WebCore/Modules/webauthn/PublicKeyCredential.cpp:
* Source/WebCore/Modules/webdatabase/SQLResultSetRowList.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/SecurityContext.cpp:
(WebCore::SecurityContext::policyContainer const):
(WebCore::SecurityContext::inheritPolicyContainerFrom):
* Source/WebCore/dom/SecurityContext.h:
(WebCore::SecurityContext::ipAddressSpace const):
(WebCore::SecurityContext::setIPAddressSpace):
* Source/WebCore/loader/LinkLoader.cpp:
* Source/WebCore/loader/PolicyContainer.h:
(WebCore::PolicyContainer::isolatedCopy const):
(WebCore::PolicyContainer::isolatedCopy):
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::targetAddressSpace const):
(WebCore::ResourceRequestBase::setTargetAddressSpace):
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::m_ipAddressSpace):
(WebCore::ResourceResponseData::isolatedCopy const):
(WebCore::ResourceResponseBase::crossThreadData const):
(WebCore::ResourceResponseBase::fromCrossThreadData):
(WebCore::ResourceResponseBase::getResponseData const):
(WTF::Persistence::Coder&lt;WebCore::ResourceResponseData&gt;::decodeForPersistence):
(WebCore::m_type): Deleted.
* Source/WebCore/platform/network/ResourceResponseBase.h:
(WebCore::ResourceResponseBase::ipAddressSpace):
(WebCore::ResourceResponseBase::setIPAddressSpace):
(WebCore::ResourceResponseData::ResourceResponseData):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/IPAddressSpaceTests.cpp: Added.
(TestWebKitAPI::TEST(IPAddressSpace, IPv4Loopback)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4PrivateAddresses)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4CarrierGradeNAT)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4LinkLocal)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4Benchmarking)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4PublicAddresses)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6Loopback)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6UniqueLocal)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6LinkLocal)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6MappedIPv4DottedDecimal)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6MappedIPv4HexNotation)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6PublicAddresses)):
(TestWebKitAPI::TEST(IPAddressSpace, HostnameAddresses)):
(TestWebKitAPI::TEST(IPAddressSpace, EdgeCasesAndMalformed)):
(TestWebKitAPI::TEST(IPAddressSpace, UtilityFunctions)):
(TestWebKitAPI::TEST(IPAddressSpace, DifferentURLSchemes)):
(TestWebKitAPI::TEST(IPAddressSpace, URLsWithPorts)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4BoundaryConditions)):

Canonical link: <a href="https://commits.webkit.org/298921@main">https://commits.webkit.org/298921@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3130ad07c4b0d5538c19806c7cb6177cc6a2f49b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116891 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36556 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27129 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122976 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/68909 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118780 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37253 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45156 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88776 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119840 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29710 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104860 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69235 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28777 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22965 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66632 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/109007 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99092 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23248 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126103 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115419 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43791 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32905 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97444 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44155 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101060 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97243 "Passed tests") | 
| [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24796 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Checked out pull request; Reviewed by Alex Christensen; Failed to compile WebKit") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42568 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20511 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40186 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18698 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43676 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49272 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144118 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43143 "Built successfully") | | [💥 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37115 "An unexpected error occured. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46482 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44848 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8bc0f5489328a1639cf0cdf7a378edb6e583864f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4a6babcd4b77468d7c45157a7c5c216bbcbe252a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bcf786bd24543d0dbf5e7e2fe3927abcdb035d67. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1a776e32b00d2be127332faf54b025df8e463d72. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

achristensen07 commented on Source/WebCore/Modules/fetch/IPAddressSpace.h:
```diff
@@ -27,9 +27,14 @@
 namespace WebCore {
 
-enum class IPAddressSpace : bool {
+enum class IPAddressSpace : uint8_t {
```
bool is sufficient to cover the 2 existing cases.

achristensen07 commented on Source/WebCore/Modules/fetch/LocalNetworkAccess.cpp:
```diff
@@ -0,0 +1,104 @@
+                WebCore::ResourceError::Type::AccessControl };
+        }
+        // Return null.
+        // return std::nullopt;
```
This doesn't look like it should stay commented out.

achristensen07 commented on Source/WebCore/platform/network/ResourceResponseBase.cpp:
```diff
@@ -940,6 +951,9 @@ void Coder<WebCore::ResourceResponseData>::encodeForPersistence(Encoder& encoder
     encoder << data.proxyName;
     encoder << data.isRangeRequested;
+#if ENABLE(LOCAL_NETWORK_ACCESS)
+    encoder << data.ipAddressSpace;
```
Does this really need to be persisted?

achristensen07 commented on Source/WebCore/platform/network/ResourceResponseBase.h:
```diff
@@ -267,6 +273,9 @@ class ResourceResponseBase {
     mutable WasPrivateRelayed m_wasPrivateRelayed : bitWidthOfWasPrivateRelayed { WasPrivateRelayed::No };
     String m_proxyName;
+#if ENABLE(LOCAL_NETWORK_ACCESS)
+    std::optional<IPAddressSpace> m_ipAddressSpace;
```
Will this commonly be nullopt?  If so, we might want to use a unique_ptr to reduce sizeof(ResourceResponse)

achristensen07 commented on Source/WebCore/platform/network/ResourceResponseBase.h:
```diff
@@ -297,6 +306,30 @@ struct ResourceResponseData {
     ResourceResponseData(ResourceResponseData&&) = default;
     ResourceResponseData& operator=(ResourceResponseData&&) = default;
+#if ENABLE(LOCAL_NETWORK_ACCESS)
+    ResourceResponseData(URL&& url, String&& mimeType, long long expectedContentLength, String&& textEncodingName, int httpStatusCode, String&& httpStatusText, String&& httpVersion, HTTPHeaderMap&& httpHeaderFields, std::optional<NetworkLoadMetrics>&& networkLoadMetrics, ResourceResponseSource source, ResourceResponseBaseType type, ResourceResponseBaseTainting tainting, bool isRedirected, UsedLeg ...
```
This is a lot of duplicate code.  It might be nicer to just add one parameter protected by the ENABLE macro.  Same with RequestData.

achristensen07 commented on Source/WebCore/platform/network/ResourceRequestBase.h:
```diff
@@ -30,6 +30,7 @@
 #include "FormData.h"
 #include "FrameLoaderTypes.h"
 #include "HTTPHeaderMap.h"
+#include "IPAddressSpace.h"
```
It would be nice to forward declare and put implementations into cpp files instead of including a header from this very commonly included header and increasing the build time.

achristensen07 commented on Source/WebCore/loader/PolicyContainer.h:
```diff
@@ -39,11 +40,19 @@ struct PolicyContainer {
 
+#if ENABLE(LOCAL_NETWORK_ACCESS)
+    PolicyContainer isolatedCopy() const & { return { contentSecurityPolicyResponseHeaders.isolatedCopy(), crossOriginEmbedderPolicy.isolatedCopy(), crossOriginOpenerPolicy.isolatedCopy(), referrerPolicy, ipAddressSpace }; }
+    PolicyContainer isolatedCopy() && { return { WTFMove(contentSecurityPolicyResponseHeaders).isolatedCopy(), WTFMove(crossOriginEmbedderPolicy).isolatedCopy(), WTFMove(crossOriginOpenerPolicy).isolatedCopy(), referrerPolicy, ipAddressSpace }; }
```
This is a lot of duplicate code.  Can we instead just protect one member with #if ENABLE?

achristensen07 commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -42777,7 +42784,7 @@
 				7B66CFD52D31387D00BD6CB5 /* IOSurfaceDrawingBuffer.h in Headers */,
 				1C21E57D183ED1FF001C289D /* IOSurfacePool.h in Headers */,
-				3C8D24EE2E25CF1A00D2A929 /* IPAddressSpace.h in Headers */,
+				3C6A556C2E273847007C1B41 /* IPAddressSpace.h in Headers */,
```
Changes like this probably shouldn't be included.

achristensen07 commented on Source/WebCore/Modules/fetch/FetchRequest.h:
```diff
@@ -91,7 +91,6 @@ class FetchRequest final : public FetchBodyOwner {
     RequestPriority priority() const { return m_priority; }
 
     const std::optional<IPAddressSpace>& targetAddressSpace() const { return m_targetAddressSpace; }
-
```
This file doesn't need to be changed in this PR.

achristensen07 reviewed (changes requested):
Needs tests.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ac1c1eb0b9b3dceb205fa4f6cae7d872957d73f6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/23dd4c74dd74f3771a11900b61cfce4f0b2c8522. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/694b788e058d44a6eb8224e1342113546e2da60b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b57cecfdc353a39dbd3619e2e288d18adf05cb34. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

stwrt commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -505,6 +505,10 @@
 #endif
 
+#if !defined(ENABLE_LOCAL_NETWORK_ACCESS)
+#define ENABLE_LOCAL_NETWORK_ACCESS 0
```
I noticed that this isn't currently being used to gate any pieces of the code.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fa4e1c1354d92a6b0597ee911fab24d9ec8892fb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8fbc4b1a182d68b70e47ec1050f437eae5bdbf36. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8fbc4b1a182d68b70e47ec1050f437eae5bdbf36. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

igower commented on Source/WebCore/Modules/fetch/FetchRequestInit.h:
```diff
@@ -51,9 +51,8 @@ struct FetchRequestInit {
-
-    bool hasMembers() const { return !method.isEmpty() || headers || body || !referrer.isEmpty() || referrerPolicy || mode || credentials || cache || redirect || !integrity.isEmpty() || keepalive || !window.isUndefined() || !signal.isUndefined() || targetAddressSpace; }
+    IPAddressSpace targetAddressSpace { IPAddressSpace::Public };
+    bool hasMembers() const { return !method.isEmpty() || headers || body || !referrer.isEmpty() || referrerPolicy || mode || credentials || cache || redirect || !integrity.isEmpty() || keepalive || !window.isUndefined() || !signal.isUndefined() || targetAddressSpace != IPAddressSpace::Public; }
```
There is no need to keep this optional, the user is not intended to set the targetAddressSpace as public, and no special logic should be taken if it is set to public by default. The hasMembers function only needs to return true if the value is set to local.

achristensen07 commented on Source/WebCore/platform/network/ResourceRequestBase.h:
```diff
@@ -87,6 +90,7 @@ class ResourceRequestBase {
             , m_didFilterLinkDecoration(didFilterLinkDecoration)
             , m_isPrivateTokenUsageByThirdPartyAllowed(isPrivateTokenUsageByThirdPartyAllowed)
             , m_wasSchemeOptimisticallyUpgraded(wasSchemeOptimisticallyUpgraded)
+                , m_targetAddressSpace(targetAddressSpace)
```
nit: indentation.

achristensen07 commented on Source/WebCore/platform/network/ResourceRequestBase.h:
```diff
@@ -247,6 +252,9 @@ class ResourceRequestBase {
     ResourceRequestRequester requester() const { return m_requestData.m_requester; }
     void setRequester(ResourceRequestRequester requester) { m_requestData.m_requester = requester; }
     
+    const IPAddressSpace& targetAddressSpace() const { return m_requestData.m_targetAddressSpace; }
```
nit: no need to return const ref

achristensen07 commented on Source/WebCore/platform/network/ResourceResponseBase.h:
```diff
@@ -149,6 +150,9 @@ class ResourceResponseBase {
     void setProxyName(String&& proxyName) { m_proxyName = WTFMove(proxyName); }
     const String& proxyName() const { return m_proxyName; }
 
+    const IPAddressSpace& ipAddressSpace() const { return m_ipAddressSpace; }
```
nit: no need for const ref.

achristensen07 commented on Source/WebCore/platform/network/ResourceResponseBase.h:
```diff
@@ -289,6 +293,10 @@ class ResourceResponseBase {
     Type m_type : bitWidthOfType { Type::Default };
+
+protected:
+    IPAddressSpace m_ipAddressSpace { IPAddressSpace::Public };
```
All the other members here have bit widths.  This should too.  No need to take 8 bits when all we need is 1.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8fa754511a6251bc00d5feb8261a656595746686. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/01ee7381d8259661d1c06a5f3da833fd2408f5b6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e55b37527b839e59e5f916fc2bd261867e1c1150. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3b83d4826ec62f38d1058f984f3fd100521549eb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48741

igower committed:
Extend Fetch to accept local connections
https://bugs.webkit.org/show_bug.cgi?id=296710
rdar://154439024

Reviewed by Alex Christensen.

Integration with fetch as per the local network access spec: (https://wicg.github.io/local-network-access/#integration-with-fetch)
Upstream commit: https://github.com/web-platform-tests/wpt/commit/c4ab993576ca17eafd7108853e112d1e97aeaacc

* LayoutTests/TestExpectations:
* LayoutTests/ipc/invalid-url-network-data-task-crash.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::updateTargetAddressSpaceIfNeeded):
(WebCore::FetchRequest::initializeWith):
(WebCore::FetchRequest::resourceRequest const):
(WebCore::FetchRequest::clone):
* Source/WebCore/Modules/fetch/FetchRequest.h:
* Source/WebCore/Modules/fetch/FetchRequest.idl:
* Source/WebCore/Modules/fetch/FetchRequestInit.h:
(WebCore::FetchRequestInit::hasMembers const):
* Source/WebCore/Modules/fetch/FetchRequestInit.idl:
* Source/WebCore/Modules/fetch/IPAddressSpace.cpp: Added.
(WebCore::determineIPAddressSpace):
(WebCore::isLocalIPAddressSpace):
* Source/WebCore/Modules/fetch/IPAddressSpace.h:
* Source/WebCore/Modules/mediasession/MediaSessionCoordinator.cpp:
* Source/WebCore/Modules/mediastream/MediaStreamTrackProcessor.cpp:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpSenderBackend.h:
* Source/WebCore/Modules/notifications/NotificationEvent.cpp:
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/Modules/webaudio/AudioNode.h:
* Source/WebCore/Modules/webauthn/PublicKeyCredential.cpp:
* Source/WebCore/Modules/webdatabase/SQLResultSetRowList.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/SecurityContext.cpp:
(WebCore::SecurityContext::policyContainer const):
(WebCore::SecurityContext::inheritPolicyContainerFrom):
* Source/WebCore/dom/SecurityContext.h:
(WebCore::SecurityContext::ipAddressSpace const):
(WebCore::SecurityContext::setIPAddressSpace):
* Source/WebCore/loader/LinkLoader.cpp:
* Source/WebCore/loader/PolicyContainer.h:
(WebCore::PolicyContainer::isolatedCopy const):
(WebCore::PolicyContainer::isolatedCopy):
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::targetAddressSpace const):
(WebCore::ResourceRequestBase::setTargetAddressSpace):
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::m_ipAddressSpace):
(WebCore::ResourceResponseData::isolatedCopy const):
(WebCore::ResourceResponseBase::crossThreadData const):
(WebCore::ResourceResponseBase::fromCrossThreadData):
(WebCore::ResourceResponseBase::getResponseData const):
(WTF::Persistence::Coder<WebCore::ResourceResponseData>::decodeForPersistence):
(WebCore::m_type): Deleted.
* Source/WebCore/platform/network/ResourceResponseBase.h:
(WebCore::ResourceResponseBase::ipAddressSpace):
(WebCore::ResourceResponseBase::setIPAddressSpace):
(WebCore::ResourceResponseData::ResourceResponseData):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/IPAddressSpaceTests.cpp: Added.
(TestWebKitAPI::TEST(IPAddressSpace, IPv4Loopback)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4PrivateAddresses)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4CarrierGradeNAT)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4LinkLocal)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4Benchmarking)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4PublicAddresses)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6Loopback)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6UniqueLocal)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6LinkLocal)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6MappedIPv4DottedDecimal)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6MappedIPv4HexNotation)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv6PublicAddresses)):
(TestWebKitAPI::TEST(IPAddressSpace, HostnameAddresses)):
(TestWebKitAPI::TEST(IPAddressSpace, EdgeCasesAndMalformed)):
(TestWebKitAPI::TEST(IPAddressSpace, UtilityFunctions)):
(TestWebKitAPI::TEST(IPAddressSpace, DifferentURLSchemes)):
(TestWebKitAPI::TEST(IPAddressSpace, URLsWithPorts)):
(TestWebKitAPI::TEST(IPAddressSpace, IPv4BoundaryConditions)):

Canonical link: https://commits.webkit.org/298921@main
